File: 백준/Gold/1202. 보석 도둑/보석 도둑.java
import java.io.*;
import java.util.*;



public class Main {

    // ! 보석 도둑
    static int n, k;
    static long weight, value; // ! 보석의 무게, 가치
    static List<Long> bags = new ArrayList<>();
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    private static class Data implements Comparable<Data>{
        long weight, value;

        Data(long weight, long value) {
            this.weight = weight;
            this.value = value;
        }

        @Override
        public int compareTo(Data data) {
            if (this.weight == data.weight) {
                Long.compare(data.value, this.value);
            }
            return Long.compare(this.weight, data.weight); // ! 무게 기준 오름차순 먼저 2. 같다면 가치 기준 내림차순 진행
        }
    }

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken()); // ! n개의 보석
        k = Integer.parseInt(st.nextToken()); // ! k개의 가방
        PriorityQueue<Data> pq = new PriorityQueue<>();
        PriorityQueue<Long> temp = new PriorityQueue<>((a,b) -> b.compareTo(a));
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            weight = Long.parseLong(st.nextToken());
            value = Long.parseLong(st.nextToken());
            pq.offer(new Data(weight, value));
        }
        for (int i = 0; i < k; i++) {
            long max_weight = Long.parseLong(br.readLine());
            bags.add(max_weight);
        }

        Collections.sort(bags); // ! 오름차순 정렬
        long res = 0;
        for (int i = 0; i < k; i++) {
            long max_weight = bags.get(i); // ! 가방의 최대 무게 가져온 후
            while (!pq.isEmpty() && pq.peek().weight <= max_weight) {
                Data data = pq.poll();
                temp.offer(data.value); // ! 후보군 등록
            }

            if (!temp.isEmpty()) { // ! 존재하면 더해야함
                res += temp.poll();
            }
        }
        System.out.println(res);
    }
}

File: 백준/Gold/1202. 보석 도둑/보석 도둑.py
import heapq
import sys
from collections import deque


# 보석 털기
# 보석 총 N개, 각 보석 무게 m과 가격 v
# k개의 가방. 각 가방에 담을 수 있는 최대 무게 c
# 가방에는 한개의 보석만 넣을 수 있다.
# 훔칠 수 있는 보석의 최대 가격 구하기.
n, k = map(int, input().split())  # 보석 개수, 가방 개수
data = []
for _ in range(n):
    m, v = map(int, input().split())
    data.append((m, v))  # 각 보석의 무게, 가격

bags = []
for _ in range(k):  # k개 가방의 최대 무게
    bags.append(int(input()))  # 가방의 최대 무게

bags.sort() # 가장 최대 무게 오름차순 정렬
data.sort(key = lambda x : x[0]) # 보석을 무게 순으로 오름차순 정렬
res = 0
heap = []
dq = deque(data)
for weight in bags: # 각 가방의 최대 무게 하나씩 꺼내서
    while dq and dq[0][0] <= weight: # 현재 가방의 최대 무게보다 작다면
        m,v = dq.popleft() # 꺼낸 보석의 무게, 가격
        heapq.heappush(heap, -v) # 보석의 가격 최대 힙으로
    if heap: # 꺼낸 보석들이 존재한다면 그 중 최대 가치 꺼내야함
        res += -heapq.heappop(heap)
print(res)